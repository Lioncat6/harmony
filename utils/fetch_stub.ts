import type { ErrorConstructor } from '@/utils/errors.ts';
import { dirname } from '@std/path/dirname';
import { stub } from '@std/testing/mock';
import { urlToFilePath } from './file_path.ts';

/** CLI flag (`--download`) which allows {@linkcode stubFetchWithCache} etc. to make network requests. */
export const downloadMode = Deno.args.includes('--download');

/**
 * Stubs {@linkcode fetch} to load the response from a cache instead of making a network request.
 *
 * Only in {@linkcode downloadMode} the resource will be fetched and written to the cache (as a file).
 */
export function stubFetchWithCache(cacheDir = 'testdata') {
	const originalFetch = globalThis.fetch;

	return stub(globalThis, 'fetch', async function (input, init) {
		const url = new URL(input instanceof Request ? input.url : input);
		const path = await urlToFilePath(url, { baseDir: cacheDir });

		if (downloadMode) {
			const response = await originalFetch(input, init);
			await saveResponse(response.clone(), path);
			return response;
		} else {
			return loadResponse(path);
		}
	});
}

/** Writes the body of the given response to the file at the given path. */
export async function saveResponse(response: Response, path: string): Promise<void> {
	const { body } = response;
	if (response.ok && body) {
		await Deno.mkdir(dirname(path), { recursive: true });
		await Deno.writeFile(path, body);
	}
}

/** Reads the cached response body from the file at the given path. */
export async function loadResponse(path: string, NotFoundError: ErrorConstructor = Error): Promise<Response> {
	try {
		const body = await Deno.readFile(path);
		return new Response(body);
	} catch (error) {
		if (error instanceof Deno.errors.NotFound) {
			throw new NotFoundError(`Response has not been cached at ${path}`);
		} else {
			throw error;
		}
	}
}
