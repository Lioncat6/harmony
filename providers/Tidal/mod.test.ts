// Automatically load .env environment variable file (before anything else).
import 'std/dotenv/load.ts';

import { describeProvider, makeProviderOptions } from '@/providers/test_spec.ts';
import { stubProviderLookups, stubTokenRetrieval } from '@/providers/test_stubs.ts';
import { downloadMode } from '@/utils/fetch_stub.ts';
import { assert } from 'std/assert/assert.ts';
import { afterAll, describe } from '@std/testing/bdd';
import type { Stub } from '@std/testing/mock';
import { assertSnapshot } from '@std/testing/snapshot';

import TidalProvider from './mod.ts';

describe('Tidal provider', () => {
	const tidal = new TidalProvider(makeProviderOptions());
	const stubs: Stub[] = [stubProviderLookups(tidal)];

	if (!downloadMode) {
		stubs.push(stubTokenRetrieval(tidal));
	}

	describeProvider(tidal, {
		urls: [{
			description: 'release page',
			url: new URL('https://tidal.com/album/357676034'),
			id: { type: 'album', id: '357676034' },
			isCanonical: true,
		}, {
			description: 'release /browse page',
			url: new URL('https://tidal.com/browse/album/130201923'),
			id: { type: 'album', id: '130201923' },
		}, {
			description: 'listen.tidal.com release page',
			url: new URL('https://listen.tidal.com/album/11343637'),
			id: { type: 'album', id: '11343637' },
		}, {
			description: 'release page with /track suffix',
			url: new URL('https://listen.tidal.com/album/301366648/track/301366649'),
			id: { type: 'album', id: '301366648' },
		}, {
			description: 'artist page',
			url: new URL('https://tidal.com/artist/80'),
			id: { type: 'artist', id: '80' },
			isCanonical: true,
		}, {
			description: 'artist /browse page',
			url: new URL('https://tidal.com/browse/artist/3557299'),
			id: { type: 'artist', id: '3557299' },
		}, {
			description: 'listen.tidal.com artist page',
			url: new URL('https://listen.tidal.com/artist/116'),
			id: { type: 'artist', id: '116' },
		}, {
			description: 'track page',
			url: new URL('https://tidal.com/browse/track/11343638'),
			id: undefined,
		}],
		releaseLookup: [{
			description: 'live album with video tracks and featured artist (v1 API)',
			release: new URL('https://tidal.com/album/130201923'),
			options: {
				// Use data from an old snapshot which was made when the v1 API was still live.
				snapshotMaxTimestamp: 1717684821,
				regions: new Set(['GB']),
			},
			assert: async (release, ctx) => {
				await assertSnapshot(ctx, release);
				const allTracks = release.media.flatMap((medium) => medium.tracklist);
				assert(allTracks[5].artists?.length === 2, 'Track 6 should have two artists');
				assert(allTracks[8].type === 'video', 'Track 9 should be a video');
				assert(allTracks.every((track) => track.isrc), 'All tracks should have an ISRC');
			},
		}],
	});

	afterAll(() => {
		stubs.forEach((stub) => stub.restore());
	});
});
