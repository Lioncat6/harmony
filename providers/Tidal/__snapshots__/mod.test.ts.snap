export const snapshot = {};

snapshot[`Tidal provider > release lookup > live album with video tracks and featured artist (v1 API) 1`] = `
{
  artists: [
    {
      creditedName: "JAY Z",
      externalIds: [
        {
          id: "7804",
          provider: "tidal",
          type: "artist",
        },
      ],
      name: "JAY Z",
    },
  ],
  copyright: "© 2012 S. Carter Enterprises, LLC.",
  externalLinks: [
    {
      types: [
        "paid streaming",
      ],
      url: "https://tidal.com/browse/album/130201923",
    },
  ],
  gtin: "00602508779602",
  images: [
    {
      thumbUrl: "https://resources.tidal.com/images/bf26cf4f/1f9a/4f70/a485/e707c3c89079/320x320.jpg",
      types: [
        "front",
      ],
      url: "https://resources.tidal.com/images/bf26cf4f/1f9a/4f70/a485/e707c3c89079/1280x1280.jpg",
    },
  ],
  info: {
    messages: [],
    providers: [
      {
        apiUrl: "https://openapi.tidal.com/albums/130201923?countryCode=GB",
        id: "130201923",
        internalName: "tidal",
        lookup: {
          method: "id",
          region: "GB",
          value: "130201923",
        },
        name: "Tidal",
        url: "https://tidal.com/album/130201923",
      },
    ],
  },
  labels: [],
  media: [
    {
      format: "Digital Media",
      number: 1,
      tracklist: [
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200001",
          length: 273000,
          number: 1,
          title: "Empire State Of Mind (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200002",
          length: 195000,
          number: 2,
          title: "Run This Town (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200003",
          length: 212000,
          number: 3,
          title: "On to the Next One (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200004",
          length: 247000,
          number: 4,
          title: "Give It To Me/Big Pimpin (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200005",
          length: 79000,
          number: 5,
          title: "03' Bonnie & Clyde (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "Beyoncé",
              externalIds: [
                {
                  id: "1566",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "Beyoncé",
            },
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200006",
          length: 187000,
          number: 6,
          title: "Crazy In Love (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200007",
          length: 152000,
          number: 7,
          title: "Public Service Announcement (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
            {
              creditedName: "Beyoncé",
              externalIds: [
                {
                  id: "1566",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "Beyoncé",
            },
          ],
          isrc: "USJZ11200008",
          length: 350000,
          number: 8,
          title: "Young Forever (Live)",
          type: "audio",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200009",
          length: 274000,
          number: 9,
          title: "Empire State Of Mind (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200010",
          length: 196000,
          number: 10,
          title: "Run This Town (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200011",
          length: 213000,
          number: 11,
          title: "On to the Next One (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200012",
          length: 248000,
          number: 12,
          title: "Give It 2 Me / Big Pimpin (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200013",
          length: 80000,
          number: 13,
          title: "03' Bonnie & Clyde (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "Beyoncé",
              externalIds: [
                {
                  id: "1566",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "Beyoncé",
            },
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200014",
          length: 188000,
          number: 14,
          title: "Crazy In Love (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
          ],
          isrc: "USJZ11200015",
          length: 153000,
          number: 15,
          title: "Public Service Announcement (Live)",
          type: "video",
        },
        {
          artists: [
            {
              creditedName: "JAY Z",
              externalIds: [
                {
                  id: "7804",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "JAY Z",
            },
            {
              creditedName: "Beyoncé",
              externalIds: [
                {
                  id: "1566",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "Beyoncé",
            },
          ],
          isrc: "USJZ11200016",
          length: 351000,
          number: 16,
          title: "Young Forever (Live)",
          type: "video",
        },
      ],
    },
  ],
  packaging: "None",
  releaseDate: {
    day: 1,
    month: 1,
    year: 2012,
  },
  status: "Official",
  title: "Live In Brooklyn",
  types: [
    "Album",
  ],
}
`;

snapshot[`Tidal provider > release lookup > lyric video (v2 API) 1`] = `
{
  artists: [
    {
      creditedName: "Taylor Swift",
      externalIds: [
        {
          id: "3557299",
          provider: "tidal",
          type: "artist",
        },
      ],
      name: "Taylor Swift",
    },
  ],
  copyright: "© 2024 Taylor Swift",
  externalLinks: [
    {
      types: [
        "paid streaming",
      ],
      url: "https://tidal.com/video/358461354",
    },
  ],
  gtin: undefined,
  images: [
    {
      thumbUrl: "https://resources.tidal.com/images/931df7cf/57ce/47f8/9a6e/c7cea3e19287/320x180.jpg",
      types: [
        "front",
      ],
      url: "https://resources.tidal.com/images/931df7cf/57ce/47f8/9a6e/c7cea3e19287/1280x720.jpg",
    },
  ],
  info: {
    messages: [],
    providers: [
      {
        apiUrl: "https://openapi.tidal.com/v2/videos?countryCode=GB&include=artists%2Citems.artists%2Cproviders&filter%5Bid%5D=358461354",
        id: "video/358461354",
        internalName: "tidal",
        lookup: {
          method: "id",
          region: "GB",
          value: "video/358461354",
        },
        name: "Tidal",
        url: "https://tidal.com/video/358461354",
      },
    ],
  },
  labels: [
    {
      name: "Universal",
    },
  ],
  media: [
    {
      format: "Digital Media",
      number: 1,
      tracklist: [
        {
          artists: [
            {
              creditedName: "Taylor Swift",
              externalIds: [
                {
                  id: "3557299",
                  provider: "tidal",
                  type: "artist",
                },
              ],
              name: "Taylor Swift",
            },
          ],
          isrc: "USUMV2400558",
          length: 208000,
          number: 1,
          recording: {
            externalIds: [
              {
                id: "358461354",
                provider: "tidal",
                type: "track",
              },
            ],
          },
          title: "My Boy Only Breaks His Favorite Toys (Lyric Video)",
          type: "video",
        },
      ],
    },
  ],
  packaging: "None",
  releaseDate: {
    day: 19,
    month: 4,
    year: 2024,
  },
  status: "Official",
  title: "My Boy Only Breaks His Favorite Toys (Lyric Video)",
  types: [
    "Single",
  ],
}
`;
